File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionHasValue.java
package org.hisp.dhis.rules.functions;

import org.hisp.dhis.rules.RuleExpression;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor;
import org.hisp.dhis.rules.parser.expression.function.ScalarFunctionToEvaluate;

import java.util.Map;

import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;
import static org.hisp.dhis.rules.parser.expression.CommonExpressionVisitor.DEFAULT_BOOLEAN_VALUE;

public class RuleFunctionHasValue
    extends ScalarFunctionToEvaluate
{
    @Override
    public Object evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        Map<String, RuleVariableValue> valueMap = visitor.getValueMap();

        String variableName = RuleExpression.getProgramRuleVariable( ctx );
        RuleVariableValue variableValue = valueMap.get( variableName );

        if ( variableValue == null )
        {
            return String.valueOf( false );
        }

        return String.valueOf( valueMap.get( variableName ).value() != null );
    }

    @Override
    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
    {
        ctx

        visitor.getItemDescriptions().put( ctx.getText(), "" );

        return DEFAULT_BOOLEAN_VALUE;
    }
}

File: src/main/java/org/hisp/dhis/rules/functions/RuleFunctionHasValue.kt
package org.hisp.dhis.rules.functions

import org.hisp.dhis.rules.RuleVariableValue


class RuleFunctionHasValue : RuleFunction() {

    override fun evaluate(arguments: List<String>,
                          valueMap: Map<String, RuleVariableValue>?, supplementaryData: Map<String, List<String>>?): String {
        when {
            valueMap == null -> throw IllegalArgumentException("valueMap is expected")
            arguments.size != 1 -> throw IllegalArgumentException("One argument was expected, ${arguments.size} were supplied")

            // ToDo: make sure that argument names are actually argument names and not values.
            else -> {
                val variableName = arguments[0].replace("'", "")
                //val variableValue = valueMap[variableName] ?: return false.toString()

                return (!valueMap[variableName]?.value().isNullOrEmpty()).toString()
            }
        }
    }

    companion object {
        const val D2_HAS_VALUE = "d2:hasValue"

        @JvmStatic
        fun create() = RuleFunctionHasValue()
    }
}
